Switch to modern JS practices

aburgel opened the pull request:
/domain @Betterment/test_track_core @HipsterBrown
/no-platform

This changeset updates the client to use some newer JS best practices. I put this together to learn about the JS ecosystem and have gotten it to a place that is ready for review.

Some of the things I did:

* Switched from grunt to rollup
* Used ES5 modules (along with babel for transpiling)
* Rewrote tests using jest
* Removed build artifacts from source control
* Tweaked some code in minor ways (mostly to ease testing)

Some of the things I intentionally did not do (mostly to keep the diff small):

* Use ES6 classes 
* Replace dependencies like jquery, jquery-cookie

I'll add some comments inline to call out interesting bits.

Fixes #6

aburgel commented on package.json:
```diff
@@ -2,9 +2,12 @@
   "description": "Javascript Client for Test Track",
-  "main": "dist/testTrack.js",
+  "main": "dist/testTrack.bundle.js",
+  "module": "dist/testTrack.js",
```
This is a new standard for exporting modules from npm packages: https://github.com/rollup/rollup/wiki/pkg.module

aburgel commented on rollup.config.js:
```diff
@@ -0,0 +1,43 @@
+import resolve from 'rollup-plugin-node-resolve';
+import { terser } from "rollup-plugin-terser";
+
+export default [
```
this rollup config produces two artifacts:

1. `dist/testTrack.js` which is an ES6 module version of the library, it treats jquery (and crypto which is provided by the browser) as external, so as not to bloat the size. The other dependencies, like uuid and md5 are bundled cuz they are specific to this library and pretty small.

2. `dist/testTrack.bundle.js` which is equivalent to the current artifact. its a UMD style JS file, that is minified and uglified. it also does not include jquery.

aburgel commented on src/abConfiguration.js:
```diff
@@ -1,66 +1,66 @@
-var ABConfiguration = (function() { // jshint ignore:line
```
most of the changes to the code is formatting, you should review with whitespace off.

aburgel commented on src/assignmentNotification.js:
```diff
@@ -1,49 +1,60 @@
 
-    return _AssignmentNotification;
-})();
+    return $.when(firstPersist, secondPersist);
```
this change is for testing. i want to know when the persist calls completes, so now it returns a promise.

aburgel commented on src/assignmentOverride.js:
```diff
@@ -1,60 +1,44 @@
-        }
-    };
-
-    _AssignmentOverride.prototype.send = function() {
```
this method is never used. i think it was copy/pasted from AssignmentNotification.

aburgel commented on src/assignmentOverride.js:
```diff
@@ -1,60 +1,44 @@
+            split_name: this._assignment.getSplitName(),
+            variant: this._assignment.getVariant(),
+            context: this._assignment.getContext(),
+            mixpanel_result: 'success' // we don't want to track overrides
```
`trackResult` arg is never used, so i'm removing and hardcoding this to success since we don't want analytics to know about overrides

aburgel commented on src/bindPolyfill.js:
```diff
@@ -1,29 +0,0 @@
-(function() {
```
i would love some feedback on this. i don't think we should be including polyfills in a library. my view is that should be the responsibility of the application (it knows what browsers are supported and can decide whether to polyfill).

should we continue to include this in the UMD artifact for backwards compatibility?

aburgel replied:
@samandmoore what's your take on this question?

samandmoore replied:
I'm fine with this choice. It's tricky because we'd wanna document that you need this polyfill in some cases otherwise folks won't know until their app blows up. So, maybe put this in the readme along with jquery, jquery cookie, and whatever else the consumer needs to BYO

aburgel commented on src/session.js:
```diff
@@ -1,136 +1,143 @@
-            visitor.vary(splitName, options);
-        });
-    };
+    return this._visitorDeferred.promise();
```
returning the promise for testing

aburgel commented on src/testTrack.js:
```diff
@@ -1,72 +1,28 @@
-    // @include ../src/session.js
-    // @include ../src/identifier.js
-    // @include ../src/varyDSL.js
-    // @include ../src/abConfiguration.js
```
all of the above is now handled by rollup

samandmoore commented on rollup.config.js:
```diff
@@ -0,0 +1,43 @@
+    },
+    plugins: [
+      resolve(),
+      commonjs(),
```
does this need to have the `commonjs` plugin since it's the `esm` targetted file?

aburgel replied:
the [plugin](https://github.com/rollup/rollup-plugin-commonjs) converts commonjs modules to es6 modules, so rollup can make use of them. its still needed here.

samandmoore commented on test/specs/abConfiguration.spec.js:
```diff
@@ -13,130 +27,130 @@ describe('ABConfiguration', function() {
     });
 
-    it('requires a splitName', function() {
+    test('requires a splitName', () => {
```
since we're using jest, can't we keep using the bdd style here with describe/it? i prefer that style.

aburgel replied:
lemme see if i can get this working

aburgel replied:
done!

samandmoore commented:
this looks good to me. minor thing but i'd also like to switch to the style of putting the test files next to the source files: `foo.js` and `foo.test.js` kinda thing.

samandmoore commented:
maybe a comment or two, but <<domainlgtm

aburgel commented:
>  i'd also like to switch to the style of putting the test files next to the source files: foo.js and foo.test.js kinda thing.

agreed. but i wanted to keep the diff small. i can do this in a separate PR.

samandmoore commented:
Yup that sounds great

On Fri, May 24, 2019, 18:02 Alex Burgel <notifications@github.com> wrote:

> i'd also like to switch to the style of putting the test files next to the
> source files: foo.js and foo.test.js kinda thing.
>
> agreed. but i wanted to keep the diff small. i can do this in a separate
> PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Betterment/test_track_js_client/pull/14?email_source=notifications&email_token=AADDQIQ77IU532MB3FIHR33PXBQX3A5CNFSM4HOFA5HKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWGU3ZY#issuecomment-495799783>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADDQITFDNTZCQ7AX5JURQTPXBQX3ANCNFSM4HOFA5HA>
> .
>

aburgel commented:
i made a couple of tweaks:

1. changed how the es6 artifact is built, it no longer contains any dependencies. after reading more about this, i couldn't see a reason for bundling them since anyone who uses this version will be using a package manager and build tool that will take care of including dependencies for them.

2. updated the readme to reflect this, along with a note about the bind polyfill.

3. changed the version number to 2.0.0-alpha-1 since without the polyfill this is no longer a drop-in replacement for 1.3.8.
